FR #982: Update Billed and Effective Cost Normative Requirements (#1966)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/billedcost.md

Reviewer: @Matt-Cowsert This is a general comment, but please address the merge conflicts in this PR.  You can resolve this thread once you've done so.

Author: Thanks @shawnalpay, resolving this one and on to the last open comment.

# Billed Cost

> Billed Cost represents the cost of a [*charge*](#glossary:charge) as invoiced by the [invoice issuer](#datasets.costandusage.invoiceissuername) in a given [*billing period*](#glossary:billing-period). Billed Cost differs from [Effective Cost](#datasets.costandusage.effectivecost) when *charges* (both prepaid and postpaid) are invoiced separately from usage.

Reviewer 2: ```suggestion
Billed Cost represents the cost of a [*charge*](#glossary:charge) as invoiced by the [invoice issuer](#datasets.costandusage.invoiceissuername) in a given [*billing period*](#glossary:billing-period). Billed Cost differs from [Effective Cost](#datasets.costandusage.effectivecost) when *charges* [*covered*](#glossary:covered-charge) and [*covering charges*](#glossary:covering-charge) are invoiced separately from usage.
```

Reviewer: @thecloudman 
Prepaid and postpaid aren't the same as covering and covered.  If anything, it's more akin to the new Contract Commitment Payment Model column we add to the Contract Commitment dataset -- but I didn't suggest adding that here.

That said, I could see finessing the final sentence if you want to add those terms into the description:

```suggestion
Billed Cost represents the cost of a [*charge*](#glossary:charge) as invoiced by the [invoice issuer](#datasets.costandusage.invoiceissuername) in a given [*billing period*](#glossary:billing-period). Billed Cost differs from Effective Cost when [covering charges](#glossary:covering-charge) (e.g., prepaid or postpaid commitment purchases) are invoiced separately from the [covered charges](#glossary:covered-charge) to which they are applied.
```

All of _that_ said, I'm fine with the way it's currently written.

Reviewer 3: Not a blocking comment, but is "invoiced" the correct term here? Invoiced means to send an invoice for.

`as invoiced by the [invoice issuer]` is correct in that the dollars for billed cost would match the dollars on the invoice.

`are invoiced separately from` is questionable in that the both the covering covering and covered charges may appear on the same invoice. I think this would be accurate if we're referring to the sum of the column, but if we're referring to the values in a single row (charge), it's not quite accurate.

Ignoring the context, "charged" would be my default verb here, but "charges are charged" isn't great. The next best suggestion I have would be "recorded"

Reviewer 3: Aside from that, I'm fine with the way it's currently written and am fine with Shawn's suggestion.

Author: Adopting a combined version that threads in covered/covering (per @thecloudman's suggestion and @shawnalpay's refinement) and replaces "invoiced separately" with "recorded separately" (per @kk09v's observation that covering and covered charges can appear on the same invoice).

Updated text:

> Billed Cost differs from Effective Cost when covering charges (e.g., prepaid or postpaid commitment purchases) are recorded separately from the covered charges to which they are applied.

Thanks, all three. This is tighter than where we started.

* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be denominated in the BillingCurrency.

> * BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.

Reviewer 4: Additional requirement agreed during TF-3 meeting (Feb 27).

```suggestion
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.
* BilledCost MUST NOT include any portion covered by related purchase *charges*.
```

Reviewer 2: ```suggestion
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.
* BilledCost MUST NOT include any portion covered by related purchase *charges* paid for or invoiced upfront.
```
Not sure if the wording is perfect but the second line, if accepted could impact billed cost related covered charges paid for monthly. As it reads i believe this line could inadvertently ignore covered charges that are paid for monthly

* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be denominated in the BillingCurrency.
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.
* BilledCost MUST NOT include any portion covered by related purchase *charges*.
* BilledCost MUST reflect the amount that's invoiced by the InvoiceIssuerName when the *charge* is originally generated.
* *Charges* with BilledCost greater than 0 MUST NOT be created by entities who are not the responsible or authorized party for invoicing the charge.

> * The sum of the BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) MUST match the sum of the payable amount provided in the corresponding invoice with the same id generated by the [InvoiceIssuerName](#datasets.costandusage.invoiceissuername).

Reviewer 4: Slightly reworded alternative

```suggestion
* The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding issued invoice.
```

* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be denominated in the BillingCurrency.
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.

> * BilledCost MUST be 0 when InvoiceIssuerName does not match the FOCUS dataset data generator.

Reviewer 4: ### Concerns

- **Proposed requirement:** `InvoiceIssuerName != DataGenerator`
  - A FinOps Tool Vendor can also be a Data Generator. Under this requirement, a FinOps Tool Vendor would be required to set `BilledCost = 0` for *all* charges in the FOCUS Cost and Usage artifacts they deliver, which is not correct.

- **[FOCUS 1.3 requirements model](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/1577-ai-add-all-13-spec-changes-to-the-requirements-model/specification/requirements_model/model_rules/datasets/cost_and_usage/columns/billedcost.json) — interpretation of:** `ServiceProviderName != InvoiceIssuerName`
  - How does this apply to purchases via a Marketplace?

- **[FOCUS 1.2 requirements model](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/requirements_model/model_rules/columns/billedcost.json) — interpretation of:** `ProviderName != InvoiceIssuerName`
  - This appears to be the best semantic fit, but `ProviderName` is now deprecated.

### Suggestion

- Consider introducing an additional participating entity to explicitly represent the **source of the charge information**, distinct from the Data Generator and Invoice Issuer.

> * BilledCost MUST be 0 when InvoiceIssuerName does not match the DataOriginatorName.

**Use cases:**

- **UC-1:** Marketplace purchase (CSP) with corresponding usage (SaaS)
- **UC-2:** Marketplace purchase (CSP) without corresponding usage (SaaS)

| Data Generator | Source | Use case | Charge Category | Billed Cost | Effective Cost | Service Provider | Invoice Issuer |
|----------------|--------|----------|-----------------|-------------|----------------|------------------|----------------|
| CSP | CSP | UC-1 | Purchase | 10 | 0 | SaaS | CSP |
| CSP | SaaS | UC-1 | Usage | 0 | 10 | SaaS | CSP |
| SaaS | CSP | UC-1 | Purchase | 10 | 0 | SaaS | CSP |
| SaaS | SaaS | UC-1 | Usage | 0 | 10 | SaaS | CSP |
| FinOps Tool | CSP | UC-1 | Purchase | 10 | 0 | SaaS | CSP |
| FinOps Tool | SaaS | UC-1 | Usage | 0 | 10 | SaaS | CSP |
| FinOps Tool | CSP | UC-1 | Purchase | 10 | 0 | SaaS | CSP |
| FinOps Tool | SaaS | UC-1 | Usage | 0 | 10 | SaaS | CSP |
| CSP | CSP | UC-2 | Purchase | 10 | 10 | SaaS | CSP |
| SaaS | CSP | UC-2 | Purchase | 10 | 10 | SaaS | CSP |
| FinOps Tool | CSP | UC-2 | Purchase | 10 | 10 | SaaS | CSP |

---

### Resources:

- Data Generator: 

> Human-readable name of the entity that generated the dataset instance. The Data Generator ensures the technical accuracy and delivery of the data.

- Existing [FOCUS 1.2 BilledCost requirements model rule](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/requirements_model/model_rules/columns/billedcost.json):

``` json
  ...
    "Type": "Static",
    "ValidationCriteria": {
      "MustSatisfy": "BilledCost MUST be 0 for charges where payments are received by a third party (e.g., marketplace transactions).",
      "Keyword": "MUST",
      "Requirement": {
        "CheckFunction": "CheckValue",
        "ColumnName": "BilledCost",
        "Value": 0
      },
      "Condition": {
        "CheckFunction": "CheckNotSameValue",
        "ColumnAName": "ProviderName",
        "ColumnBName": "InvoiceIssuerName"
      },
      "Dependencies": [
        "ProviderName-C-000-M",
        "InvoiceIssuerName-C-000-M"
      ]
    }
  },
  ...
```

ProviderName column is deprecated in FOCUS 1.3 and here is the current suggestion for [FOCUS 1.3 BilledCost requirements model rule](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/1577-ai-add-all-13-spec-changes-to-the-requirements-model/specification/requirements_model/model_rules/datasets/cost_and_usage/columns/billedcost.json) update:

``` json
    ...
    "DatasetId": "CostAndUsage",
    "DatasetName": "Cost and Usage",
    "Type": "Static",
    "ValidationCriteria": {
      "MustSatisfy": "BilledCost MUST be 0 for charges where payments are received by a third party (e.g., marketplace transactions).",
      "Keyword": "MUST",
      "Requirement": {
        "CheckFunction": "CheckValue",
        "ColumnName": "BilledCost",
        "Value": 0
      },
      "Condition": {
        "CheckFunction": "CheckNotSameValue",
        "ColumnAName": "ServiceProviderName",
        "ColumnBName": "InvoiceIssuerName"
      },
      "Dependencies": [
        "CAU-InvoiceIssuerName-C-000-M",
        "CAU-ServiceProviderName-C-000-M"
      ]
    }
  },

```

Reviewer 5: @ijurica We have called this out multiple times, we do not support Finops vendors declaring they are delivering the data. Data Generator specifically uses the word Generator because they are the genesis of the data. Supporting a Finops vendor declaring the progeny of the data is something that I believe that we should support but using Data Generator is not the right approach. If we do ,then the use case of identifying the origination of the data is lost. 

We need to separate the use case of data lineage into its own feature request and not muddy the waters by trying to fit it into something that will damage an existing supported feature.

Reviewer 4: Each time this topic comes up, I find myself returning to the same core point: regardless of what we may have originally intended, what ultimately matters is **what we wrote in the specification**.

If we truly want to redefine Data Generator to mean only the original source (i.e., the genesis of the raw billing data), then we would need to change:

- the definition in the FOCUS specification, and
- the corresponding content on the FOCUS and FinOps websites.

(*Please see the attached **Resources** below — FOCUS specification, focus.finops.org, and finops.org*)

---

## Recommendation

That said, my recommendation would be that we avoid narrowing the definition. As it stands today, the responsibilities and metadata expectations associated with a Data Generator appear applicable to a FinOps vendor generating and delivering a FOCUS-compliant dataset instance.

A dataset instance may aggregate data from multiple upstream sources, but:

- each dataset instance has exactly one final entity in the chain that generated that specific dataset instance (a dataset-level metadata attribute), and
- each individual record still has exactly one original source (which could be represented as a column in the dataset).

We do not need to represent every transformation step or every entity in the transformation chain. In my view, we only need:

- the first entity in the chain (at the record level), and
- the last entity in the chain (at the dataset instance level, in metadata).

---

## Resources

### Data Generator

Source: FOCUS Specification

> **Data Generator:** Human-readable name of the entity that generated the dataset instance. The Data Generator ensures the technical accuracy and delivery of the data.

### What is FOCUS?

Source: https://focus.finops.org/what-is-focus/

> #### Billing data generators
>
> Billing data generators include, but are not limited to:
> 
> - (CSPs) a.k.a. Infrastructure-as-a-Service (IaaS) providers
> - FinOps tooling vendors and service providers
> - Software-as-a-Service (SaaS) and Platform-as-a-Service (PaaS) vendors
> - Independent Software Vendors (ISVs)
> - Private cloud vendors
> - Internal teams who do chargebacks for data center usage

### Benefits for FinOps tool vendors

Source: https://www.finops.org/insights/focus-1-0-available/

> **Vendors who offer FinOps Tools** that are designed to help with the practice of FinOps are both **consumers and generators of cloud billing data**. 
> ...
> <img width="1536" height="488" alt="image" src="https://github.com/user-attachments/assets/61025f67-ad94-4971-b141-1ab7cf7abaa4" />
> 
> *Figure 1: To understand your role with respect to FOCUS, consider whether you generate or consume cloud billing data.*

Author: The DataGenerator comparison was removed from this PR in earlier commits. The current approach uses the entity invoicing prohibition instead, which sidesteps the FinOps Tool Vendor edge case. The broader Data Generator definition question (Riley's point about genesis vs. delivery) should be tracked as a separate FR. Resolving this thread.

> * BilledCost MUST be 0 when InvoiceIssuerName does not match the FOCUS dataset data generator.

Reviewer 6: I like Irena's suggestion. My suggestion is only a temporary improvement until we have a DataOriginator or equivalent available in a future release.

```suggestion
* BilledCost MUST reflect the amount that's invoiced by the InvoiceIssuerName when the *charge* is originally generated. 
* *Charges* with BilledCost greater than 0 MUST NOT be created by entities who are not the responsible or authorized party for invoicing the charge.
```

> * BilledCost MUST be 0 when InvoiceIssuerName does not match the FOCUS dataset data generator.

Author: Per TF3 discussion (AI #1970): comparing against "FOCUS dataset data generator" breaks when a FinOps Tool Vendor is the data generator (BilledCost would incorrectly be forced to 0 for all charges). The FOCUS 1.2 requirements model used `ProviderName != InvoiceIssuerName`, and `ServiceProviderName` is the 1.3 equivalent.

@ijurica -- posting this suggestion based on the TF3 discussion. Please review and accept/reject.

```suggestion
* BilledCost MUST be 0 when [ServiceProviderName](#datasets.costandusage.serviceprovidername) does not match [InvoiceIssuerName](#datasets.costandusage.invoiceissuername).
```

Reviewer 4: Unfortunately, this doesn't hold for Marketplace scenarios (see the table above and this discussion in [PR #1668](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1668#discussion_r2778277526).

For instance in a Marketplace purchase (originated from the CSP) with corresponding usage (originated from the SaaS provider):

- ServiceProviderName ≠ InvoiceIssuerName
  - ServiceProviderName = SaaS
  - InvoiceIssuerName = CSP

- The purchase charge is invoiced by the CSP and the BilledCost will probably be non-zero.

This suggestion requires BilledCost to be 0, which contradicts the Marketplace scenario.

> * BilledCost MUST be 0 when InvoiceIssuerName does not match the FOCUS dataset data generator.

Author: During TF-3, we debated if this requirement is needed after adjusting the requirements for the effective cost scenario for marketplace charges.

> * The sum of the BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) MUST match the sum of the payable amount provided in the corresponding invoice with the same id generated by the [InvoiceIssuerName](#datasets.costandusage.invoiceissuername).

Reviewer 4: FYI, InvoiceDetail-related PR #1900

Reqs candidates:

> * The sum of InvoiceDetail.BilledCost for a given InvoiceDetail.InvoiceDetailId, InvoiceDetail.InvoiceId, and InvoiceDetail.InvoiceIssuerName MUST match the payable amount provided in the corresponding invoice line items when InvoiceDetail.InvoiceStatus is "Closed".
> * The sum of InvoiceDetail.BilledCost for a given InvoiceDetail.InvoiceDetailId, InvoiceDetail.InvoiceId, and InvoiceDetail.InvoiceIssuerName MUST match the sum of CostAndUsage.BilledCost for the same CostAndUsage.InvoiceDetailId, CostAndUsage.InvoiceId, and CostAndUsage.InvoiceIssuerName when InvoiceDetail.InvoiceStatus is "Closed".
> * The sum of InvoiceDetail.BilledCost for a given InvoiceDetail.InvoiceDetailId, InvoiceDetail.InvoiceId, and InvoiceDetail.InvoiceIssuerName MAY differ from the sum of CostAndUsage.BilledCost for the same CostAndUsage.InvoiceDetailId, CostAndUsage.InvoiceId, and CostAndUsage.InvoiceIssuerName when InvoiceDetail.InvoiceStatus is "Open".

Reviewer 4: Resolving this with the intention to add an updated version

* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be denominated in the BillingCurrency.
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.

> * BilledCost MUST NOT include any portion covered by related purchase *charges*.

Reviewer: Couple thoughts here.

* I'd prefer this to read as a MUST.  (Personal preference.  But MUST exclude reads cleaner to me than MUST NOT include.)
* It would be great to leverage the new glossary entries we've got.  Without them, I (as the lay reader) don't immediately know what "covered" means.

What do you think?

```suggestion
* BilledCost MUST exclude any portion of consumption that is offset by a [*covering charge*](#glossary:covering-charge), resulting in a BilledCost of 0 for fully [*covered charges*](#glossary:covered-charge).
```

If that's too wordy, how about:

```suggestion
* BilledCost MUST exclude any portion of consumption that is offset by a [*covering charge*](#glossary:covering-charge).
```

Or, another small tweak from there:

```suggestion
* BilledCost MUST exclude any portion of consumption that is [*covered*](#glossary:covered-charge] by a [*covering charge*](#glossary:covering-charge).
```

Author: Thanks for this, Shawn. I agree on incorporating the glossary terms — that makes the requirement self-documenting.

On the stylistic shift from MUST NOT to MUST exclude: I'd prefer to keep the current phrasing for now. Irena and I discussed this and landed in the same place you did (MUST exclude reads cleaner), but changing it here in isolation creates an inconsistency across the spec. We're planning to raise this in TF3 as a candidate for the 1.4 consistency review or as an editorial item in 1.5.

For this PR, here's a revision that adds the glossary links while keeping the existing verb form:

```suggestion
* BilledCost MUST NOT include any portion of a [*covered charge*](#glossary:covered-charge) that is offset by a [*covering charge*](#glossary:covering-charge), resulting in a BilledCost of 0 for fully *covered charges*.
```

Let me know if that works for you.

Reviewer: I don't disagree (which is to say, I agree) that we use "MUST NOT include" several times and do not use "MUST exclude".  However, that's not a guideline, and generally that usage applies to dimensions, not metrics, so I don't personally see it as an inconsistency.

At the risk of repeating myself: making changes to normative statements are not in scope for consistency review.   `MUST NOT` to `MUST` is a material change, even if it feels like just a logical opposite.   So this is something we either solve now, or we accept delaying to a future release.

I'm fine with the above.

> * BilledCost MUST NOT include any portion of a [*covered charge*](#glossary:covered-charge) that is offset by a [*covering charge*](#glossary:covering-charge), resulting in a BilledCost of 0 for fully *covered charges*.

Reviewer 2: ```suggestion
* BilledCost MUST NOT include any portion of a *covered charge* that is offset by a *covering charge*.
* BilledCost MUST be 0 for *charges* that are fully *covered* by one or more *covering charges*.
```
Current format may create ambiguity on the requirement. Splitting them creates cleaner reading and guidance.

Author: Both changes from this suggestion are already in the current text (lines 19-20). The covered charge requirement was split into two bullets and uses the glossary terms. Resolving -- please update your review when you get a chance. Thanks for the feedback.

> * BilledCost MUST be based on amounts invoiced by the [InvoiceIssuerName](#datasets.costandusage.invoiceissuername), not estimated or inferred values.

Author: ```suggestion
* BilledCost MUST reflect amounts as invoiced by the [InvoiceIssuerName](#datasets.costandusage.invoiceissuername), not estimated or inferred values.
```

> * Entities that are not responsible or authorized for invoicing this *charge* MUST NOT generate *charges* with non-zero BilledCost to avoid double-counting when merging multiple datasets.

Author: Proposing we drop this requirement, consistent with the parallel decision on EffectiveCost line 25.

**Rationale:**

- Line 20 already anchors BilledCost to "amounts as invoiced by the InvoiceIssuerName." An entity that isn't the invoice issuer can't satisfy line 20 for non-zero BilledCost, making this requirement redundant.
- Like the EffectiveCost counterpart, this prohibition is unenforceable without a column that distinguishes data originators from data processors/re-deliverers.
- Dropping both entity prohibitions together keeps the two files consistent and avoids a partial fix.

If the group agrees, I'll apply this in the same pass as the EffectiveCost line 25 removal.

```suggestion
```

Author: We talked about this during TF-3, on 2026-03-13, and we've kept this requirement  and reintroduced a parallel requirement into effective cost.

Reviewer 2: Sounds counter intuitive. Why would we reintroduce something that is redundant.

Author: We discussed this in TF-3 on 3/13 and decided to keep it (and reintroduce a parallel version into EffectiveCost). The rationale:

1. Line 20 anchors BilledCost to invoiced amounts, but it doesn't explicitly prohibit a non-invoicing entity from generating charges. An entity could argue it's "reflecting" amounts without being the invoice issuer. The prohibition makes that guidance unambiguous.

2. Udam's point: even if it's not strictly enforceable today (no DataOriginator column), it signals to providers that generating charges you're not responsible for is out of bounds. That matters for marketplace scenarios where CSPs or SaaS providers might be tempted to fabricate completeness.

3. We want symmetry between BilledCost and EffectiveCost on this protection. Dropping one but not the other creates an inconsistency that's harder to explain than the redundancy.

The long-term plan is a dataset-level requirement (once we have a column that distinguishes data originators from processors). For now, keeping the per-column prohibition is the group's preferred bridge.

> * Entities that are not responsible or authorized for invoicing this *charge* MUST NOT generate *charges* with non-zero BilledCost to avoid double-counting when merging multiple datasets.

Reviewer 2: ```suggestion
* Entities that are not responsible or authorized for invoicing a *charge* MUST NOT generate *charges* with non-zero BilledCost to avoid double-counting when merging multiple datasets.
```

Reviewer 2: this charge" implies row-level context which could be ambiguous when reading at the schema level. "a charge" is cleaner.

Author: Already addressed -- the current text on line 21 uses "a *charge*" rather than "this charge." Thanks for flagging.

> * The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding issued invoice.

Reviewer: We've been talking about this requirement a LOT in TF-1 (perhaps @ijurica has already filled you in).

TL;DR: BilledCost in both CostAndUsage will almost never _exactly_ match the values printed on an invoice (and reflected in InvoiceDetail), due to small rounding errors due to a stepdown in scale (say, scale of 6 in CostAndUsage vs scale of 2 in InvoiceDetail).

I recommend you read through the [BilledCost definition](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1900/files#diff-4feedf3b866f48dbf233870b74bcdf95527641af7ac681bdae58365d1d447808) in PR #1900.  There, you will see a tolerance calculation, which I will duplicate below for convenience.  Any aggregated value of BilledCost for a given InvoiceIssuerName/InvoiceId then has to be underneath that tolerance calc, otherwise it fails the requirement.  We have good buy-in on this calc from the TF-1, and also notably @mike-finopsorg and @pydi-aws.  

PR #1900 is on track to getting approved tomorrow.  Perhaps once it's approved, you merge commit from `working_draft` and integrate similar language here?

### Tolerance Formula

The tolerance used when comparing aggregated BilledCost values is defined as:

Tolerance = `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)`

Where:

* **Rows** — The number of [CostAndUsage](#datasets.costandusage) rows included in the aggregation for the relevant [CostAndUsage.InvoiceId](#datasets.costandusage.invoiceid) and [CostAndUsage.InvoiceIssuerName](#datasets.costandusage.invoiceissuername).
* **Subunit** — The numeric value of the smallest subunit of [BillingCurrency](#datasets.invoicedetail.billingcurrency) (for example, `0.01` for USD or `1` for JPY).

The tolerance is the **greater** of the following values:

* **100 × Subunit** — Establishes a fixed minimum tolerance equal to 100 units of the smallest currency subunit.
* **(SQRT(Rows) × 0.5) × Subunit** — Provides a tolerance that increases with the square root of the number of rows to account for rounding accumulation in larger datasets.

Reviewer 6: I disagree with this approach. It introduces a completely open and unaccounted error buffer based on the dataset size that ultimately may or may not be related to the rounding errors it's seeking to account for. FWIW, FOCUS already supports the ChargeCategory (and subcategory) value to account for rounding errors. 

I agree that the equality condition can be adjusted slightly to account for rounding errors, given that invoiced charges are typically rounded to two decimal places-- the minimum legally payable amount. Introducing a generic error buffer in the spec to account for discrepancies seems like a bad idea to me (especially since most organizations have their own determinations of what constitutes a material discrepancy during reconciliation). I'll provide a suggestion for consideration.

Reviewer: 1) This tolerance calculation was approved and merged by the Members via PR #1900 yesterday, as scheduled per my previous comment.  I welcome your suggestion, but it will have to either align with, or replace, this requirement.
2) In order to increase your confidence in the approach, in [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1900/files#r2924796275), @pydi-aws says:

> I pressure-tested the tolerance formula against a AWS payer account with ~1.97 billion CUR rows across 30 invoices covering subscription, usage, and marketplace charge type.
> 
> Result: all 30 invoices pass.
> 
> The SQRT-based formula with the $1.00 floor works well with the floor catches all subscription invoices (low row counts, hundreds of rows), while the SQRT scaling provides appropriate room for high-volume usage invoices. Also, validated that the formula correctly fails when real data gaps exist confirming it distinguishes legitimate rounding drift from actual data issues. We should be good to close this thread.

Reviewer 5: @shawnalpay don't invoice tolerances have to do with variance between purchase orders and use receipts. Do we have a problem were we are talking about comparison to use receipts and not purchase orders? Do we have a potential conflict?

Reviewer 3: I applied this formula to real data (at an extreme) and I found it to still be reasonable as an upper tolerance. 

Certainly the goal is to minimize rounding errors, but this provides an limit that is no longer subjective, so I would support this formula as long as no other errors are identified. Once we have a limit of what is allowable, now it's easier to say with certainty that something is wrong and needs to be investigated.

I'm trying to think of other additional checks that could be performed to further narrow this:
* Per row the limit would be less than half of a sub-unit (whatever it's rounded to).
* Statistically, on average, your error term should be 0. I was thinking some larger check (e.g. across periods) might help, but I'm not sure if floating point math errors would skew this since they're always expressed the same way. e.g. 0.1 =
0.1000000000000000055511151231257827021181583404541015625

Reviewer: @rileyjenk 
I had to put this thread into Gemini to understand your point; wasn't immediately groking it.  😅 

I believe you're talking about **Procurement Tolerance** (specifically, two-way or three-way matching in accounting, aka the allowable variance between a Purchase Order (PO), the receipt of goods/services, and the final invoice), whereas the rest of the thread is discussing **Computational Rounding Tolerance** (data engineering and floating-point math).  Do you agree with that assessment?

If so: the `Implementation Context` section in [InvoiceDetail.BilledCost](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/invoice_detail/columns/billedcost.md) makes clear that we're discussing/addressing the latter.

Reviewer: My first thought here was to **cut the strict equality requirement entirely** from `CostAndUsage` and replace it with a pointer to `InvoiceDetail` in an Implementation Guidance section. Something like:

> **Reconciling with Invoiced Amounts**
> Due to the high volume of rows and floating-point precision differences, the sum of `BilledCost` in this dataset may not perfectly match the exact payable amount on an issued invoice. For normative requirements and statistical tolerance formulas regarding invoice reconciliation, refer to the [InvoiceDetail.BilledCost](#datasets.invoicedetail.billedcost) specification.

This would keep `InvoiceDetail` as the single source of truth for the bill, explicitly acknowledges the rounding point, and keeps the spec clean. 

**But of course we can't just delete the requirement from `CostAndUsage` and point to `InvoiceDetail`**, because some data generators might only emit `CostAndUsage`. If they don't emit `InvoiceDetail`, we lose the normative tie to the actual bill. 

Furthermore, we need to account for open billing periods where an invoice hasn't been finalized yet. I propose we update the `CostAndUsage.BilledCost` requirement to mirror the reality of the tolerance formula when comparing against the actual issued invoice, while explicitly allowing variance for unissued invoices.

```suggestion
* When a corresponding invoice has been issued, the sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST NOT differ from the payable amount provided on that invoice by more than `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)` as defined in the Rounding Variance Tolerance formula.
* When a corresponding invoice has not yet been issued, the sum of BilledCost MAY differ from preliminary or estimated invoiced amounts.
```

We would then include the same Implementation Guidance block for the formula here in `CostAndUsage` as we did in `InvoiceDetail`. This ensures that even if a provider doesn't generate `InvoiceDetail`, their `CostAndUsage` data is still held to a specific, mathematically sound standard of accuracy against the real-world bill.

Let me know what you think.

Author: Closing the loop on this thread.

**Resolution:**
- Tolerance formula is now in a standalone [Rounding Variance Tolerance](#appendix.roundingvariancetolerance) appendix entry, per agreement in Mar 23 Maintainers.
- CostAndUsage.BilledCost line 23 uses the tolerance (MUST NOT differ by more than the tolerance when the invoice has been issued).
- InvoiceDetail.BilledCost cross-dataset comparisons reference the same tolerance.

The approach Shawn floated on Mar 19 (cut strict equality from C&U, add Implementation Guidance pointer) was considered, but TF-3 preferred keeping the normative requirement with the tolerance built in. The current text reflects that decision.

> * The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding issued invoice.

Reviewer 6: @shawnalpay @ijurica @Matt-Cowsert @rileyjenk  could this work?

```suggestion
* The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding invoice, when rows with [ChargeCategory](#datasets.costandusage.chargecategory) 'Adjustment' accounting for rounding errors are included in the dataset.
* The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding invoice, within rounding adjustments necessary to express charges in the smallest denominated unit of the BillingCurrency, when rows with [ChargeCategory](#datasets.costandusage.chargecategory) 'Adjustment' accounting for rounding errors are not included in the dataset.
```

Reviewer 4: I don't think we can count on adjustments. 

But, considering we now have InvoiceDetailId maybe we can simplify this second requirement by including the InvoiceDetailId. In that case it comes down to rounding error, right?

> * The sum of BilledCost for a given InvoiceDetailId, InvoiceId, and 
InvoiceIssuerName MUST match the payable amount provided in the corresponding 
entries on the issued invoice, within the rounding precision of the BillingCurrency

Reviewer 6: Good to go with the group's decision here.

Author: Resolving this as I've applied @shawnalpay suggestion on this same line. Fine to discuss further during TF-3 and Maintainers, but don't feel the need to keep this conversation open.

> * The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding issued invoice.

Reviewer 2: InvoiceId is unique so why is there a need to say "invoiceId and InvoiceIssuerName" it would be easier just to say InvoiceId must match the amount payable. One invoice would not split payments to multiple issuers right.

Reviewer 6: Across many merged FOCUS datasets, invoices cannot be guaranteed to be unique.

Reviewer: Discussed in Mar 19 Members and agreed we can resolve.

* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be denominated in the BillingCurrency.
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.
* BilledCost MUST NOT include any portion covered by related purchase *charges*.
* BilledCost MUST reflect the amount that's invoiced by the InvoiceIssuerName when the *charge* is originally generated.

> * *Charges* with BilledCost greater than 0 MUST NOT be created by entities who are not the responsible or authorized party for invoicing the charge.

Reviewer 4: Suggestion by @udam-stitcher:

```suggestion
* Entities that are not responsible or authorized for invoicing this charge MUST NOT generate *charges* with BilledCost > 0 to avoid double-counting when merging multiple datasets.
```

Reviewer 4: I suggest we use `non-zero BilledCost` instead of `BilledCost > 0`

Author: Accepted in 01d5c09 with Irena's `non-zero` tweak.

> * The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername) MUST match the payable amount provided on the corresponding issued invoice.

Reviewer 4: FYI:

* PR #1900 (still under review). introduces a similar requirement in InvoiceDetail.BilledCost:
  * *The sum of BilledCost for a given [InvoiceDetailId](#datasets.invoicedetail.invoicedetailid), [InvoiceId](#datasets.invoicedetail.invoiceid), and [InvoiceIssuerName](#datasets.invoicedetail.invoiceissuername) MUST match the payable amount provided in the corresponding entries on the issued invoice when [InvoiceIssueStatus](#datasets.invoicedetail.invoiceissuestatus) is "Issued".*
* If the PR #1900 is approved before this PR, note that it introduces an additional column InvoiceDetailId in CostAndUsage. The conditions in this requirement may need to be updated to include InvoiceDetailId alongside InvoiceId and InvoiceIssuerName for greater precision.
* As part of PR #1501, also still under review, we plan to add an Issued invoice glossary entry.

Author: Noted. Will update the sum-check condition to include InvoiceDetailId if PR #1900 merges first.

* BilledCost MUST conform to [NumericFormat](#attributes.numericformat) requirements.
* BilledCost MUST NOT be null.
* BilledCost MUST be a valid decimal value.
* BilledCost MUST be denominated in the BillingCurrency.
* BilledCost MUST reflect all applicable pricing adjustments, including but not limited to *negotiated discounts*, *commitment discounts*, and other applicable discount programs.
* BilledCost MUST NOT include any portion of a [*covered charge*](#glossary:covered-charge) that is offset by a [*covering charge*](#glossary:covering-charge).

> * BilledCost MUST be 0 for *charges* that are fully *covered* by one or more *covering charges*.

Reviewer 2: ```suggestion
* BilledCost MUST be 0 for usage *charges* that are fully *covered* by one or more *covering charges*.
```

Author: Keeping the current text without the "usage" qualifier. The requirement should apply to any charge that is fully covered, regardless of ChargeCategory. Narrowing to "usage" would create a gap if a Purchase charge were ever fully covered by another covering charge.

* BilledCost MUST reflect amounts as invoiced by the [InvoiceIssuerName](#datasets.costandusage.invoiceissuername), not estimated or inferred values.

> * BilledCost MUST be 0 for charges generated by entities that are not responsible or authorized for invoicing, to avoid double-counting when merging multiple datasets.

Reviewer 2: Is it necessary to have this - to avoid double-counting when merging multiple datasets.

Author: Keeping the rationale text. The entity prohibition is not intuitive on its own — "to avoid double-counting when merging multiple datasets" explains the multi-dataset context that motivates the rule. Without it, implementers may not understand why the restriction exists.

> * Entities that are not responsible or authorized for invoicing a *charge* MUST NOT generate *charges* with non-zero BilledCost to avoid double-counting when merging multiple datasets.

Reviewer 4: This version is more aligned with the normative requirements guidelines. I hope I’ve preserved the original intent. Please check 😄 

```suggestion
* BilledCost MUST be 0 for charges generated by entities that are not responsible or authorized for invoicing, to avoid double-counting when merging multiple datasets.
```

Author: Incorporated. Thanks, Irena.

> * When a corresponding invoice has been issued, the sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and InvoiceIssuerName MUST NOT differ from the payable amount provided on that invoice by more than the [Rounding Variance Tolerance](#datasets.costandusage.billedcost.implementationguidance.roundingvariancetolerance).

Reviewer 4: Same as before... aligning with the normative requirements guidelines. I hope I’ve preserved the original intent. Please check 😄

```suggestion
* The sum of BilledCost for a given [InvoiceId](#datasets.costandusage.invoiceid) and InvoiceIssuerName MUST NOT differ from the payable amount provided on the corresponding invoice by more than the [Rounding Variance Tolerance](#datasets.costandusage.billedcost.implementationguidance.roundingvariancetolerance) when the corresponding invoice has been issued.
```

Author: Incorporated — tolerance-based approach with appendix reference.

> * When a corresponding invoice has not yet been issued, the sum of BilledCost MAY differ from preliminary or estimated invoiced amounts.

Reviewer 4: ```suggestion
* The sum of BilledCost MAY differ from preliminary or estimated invoiced amounts when the corresponding invoice has not yet been issued.
```

Author: Incorporated as-is.

## Implementation Guidance

> ### Handling Rounding Discrepancies

Reviewer: When I authored this tolerance calc per #1900, I embedded it in InvoiceDetail.BilledCost because it was the only place we were using it.  If we're going to copy/paste this wholesale here in CostAndUsage.BilledCost, then we need to move this into its own section.  

It appears this was already discussed in TF-3 per AI #2092, and according to that AI, we will discuss this in Mar 23 Maintainers.

I could go either way as to whether it's an attribute or appendix.  Pros and cons.  Interested to hear what others think.

Reviewer: Per discussion and agreement in Mar 23 Maintainers, I have now moved this into a new appendix entry in this PR via a direct commit.  Resolving!

### Rounding Variance Tolerance

The tolerance used when comparing the sum of BilledCost against the payable amount on an issued invoice is defined as:

Tolerance = `MAX(100 × Subunit, (SQRT(Rows) × 0.5) × Subunit)`

Where:

* **Rows** -- The number of cost and usage rows included in the aggregation for the relevant [InvoiceId](#datasets.costandusage.invoiceid) and [InvoiceIssuerName](#datasets.costandusage.invoiceissuername).
* **Subunit** -- The numeric value of the smallest subunit of [BillingCurrency](#datasets.costandusage.billingcurrency) (for example, `0.01` for USD or `1` for JPY).

The tolerance is the **greater** of the following values:

* **100 × Subunit** -- Establishes a fixed minimum tolerance equal to 100 units of the smallest currency subunit.
* **(SQRT(Rows) × 0.5) × Subunit** -- Provides a tolerance that increases with the square root of the number of rows to account for rounding accumulation in larger datasets.

> ### Scenario 1: Small Invoice (Pass)

Reviewer 2: Should scenario's be in a supporting content file instead of in the column details file? I undersand he reasoning for the scenarios to provide context to the tolerance but following consistency this may be bette in a supporting file.

Reviewer: @thecloudman 
Yeah, I logged [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1966/files#r2969854519) on the same point, and TF-3 logged AI #2092, now that we would have this same verbiage in two columns, exactly the same.

We'll be discussing this at Maintainers tomorrow, then figure out how to address.
